Framework/HistogramData/test/HistogramRebinTest.h

```diff
@@ -0,0 +1,439 @@
+#ifndef MANTID_HISTOGRAMDATA_HISTOGRAMREBINTEST_H_
+#define MANTID_HISTOGRAMDATA_HISTOGRAMREBINTEST_H_
+
+#include <cxxtest/TestSuite.h>
+
+#include "MantidHistogramData/Histogram.h"
+#include "MantidHistogramData/HistogramRebin.h"
+#include "MantidHistogramData/LinearGenerator.h"
+
+#include <algorithm>
+#include <random>
+
+using namespace Mantid::HistogramData;
+
+class HistogramRebinTest : public CxxTest::TestSuite {
+public:
+  // This pair of boilerplate methods prevent the suite being created statically
+  // This means the constructor isn't called when running other tests
+  static HistogramRebinTest *createSuite() { return new HistogramRebinTest(); }
+  static void destroySuite(HistogramRebinTest *suite) { delete suite; }
+
+  void testExecrebin() {
+    TS_ASSERT_THROWS_NOTHING(
+        rebin(getCountsHistogram(), BinEdges(10, LinearGenerator(0, 0.5))));
+  }
+
+  void testExecRebinFrequency() {
+    TS_ASSERT_THROWS_NOTHING(
+        rebin(getFrequencyHistogram(), BinEdges(10, LinearGenerator(0, 0.5))));
+  }
+
+  void testRebinNoYModeDefined() {
+    BinEdges edges(5, LinearGenerator(0, 2));
+    // X Mode Counts
+    TS_ASSERT_THROWS(
+        rebin(Histogram(Histogram::XMode::Points, Histogram::YMode::Counts),
+              edges),
+        std::runtime_error);
+    // No YMode set
+    TS_ASSERT_THROWS(
+        rebin(Histogram(BinEdges(10, LinearGenerator(0, 0.5))), edges),
+        std::runtime_error);
+  }
+
+  void testRebinFailsBinEdgesInvalid() {
+    std::vector<double> binEdges{1, 2, 3, 3, 5, 7};
+    BinEdges edges(binEdges);
+
+    TS_ASSERT_THROWS(rebin(getCountsHistogram(), edges), std::runtime_error);
```

Comment:
Maybe just add a few more cases, e.g., with the bad values at start and end?

We should also test with the new bad bin edges coinciding with the old once *and* not coinciding. Remember that this was an issue with the old implementation?